internal/x/phab/phab.go

```
@@ -0,0 +1,79 @@
+// THE SOFTWARE.
+
+// Package phab provides functionality to interact with Phabricator.
+//
+// The primary purpose is to convert failures to JSON that is compatible
+// with the Harbormaster API.
+//
+// https://secure.phabricator.com/conduit/method/harbormaster.sendmessage
+package phab
+
+import "github.com/uber/prototool/internal/x/text"
+
+const (
+	// DefaultHarbormasterLintResultName is the default name
+	// used when populating a HarbormasterLintResult.
+	DefaultHarbormasterLintResultName = "PROTOTOOL"
+	// DefaultHarbormasterLintResultCode is the default code
+	// used when populating a HarbormasterLintResult.
+	DefaultHarbormasterLintResultCode = "PROTOTOOL"
+	// DefaultHarbormasterLintResultSeverity is the default severity
+	// used when populating a HarbormasterLintResult.
+	DefaultHarbormasterLintResultSeverity = "error"
+)
+
+// HarbormasterLintResult represents a text.Failure in a structure
+// compatible with a Harbormaster Lint Result. It is meant to be
+// encoded to JSON.
+//
+// https://secure.phabricator.com/conduit/method/harbormaster.sendmessage
+type HarbormasterLintResult struct {
+	Name        string `json:"name,omitempty"`
+	Code        string `json:"code,omitempty"`
+	Severity    string `json:"severity,omitempty"`
+	Path        string `json:"path,omitempty"`
+	Line        int    `json:"line,omitempty"`
+	Char        int    `json:"char,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+// TextFailureToHarbormasterLintResult converts a text.Failure to a HarbormasterLintResult.
+func TextFailureToHarbormasterLintResult(textFailure *text.Failure) *HarbormasterLintResult {
+	if textFailure == nil {
+		return nil
+	}
+	harbormasterLintResult := &HarbormasterLintResult{
+		Name:     DefaultHarbormasterLintResultName,
+		Code:     textFailure.ID,
+		Severity: DefaultHarbormasterLintResultSeverity,
+		// TODO: Filename is not a required field on text.Failure
```

Should we return an error if the `text.Failure` doesn't have a Filename? If we NEED the Path, it might be best to handle that here.

---

Yea fair point. Will update.

---

+1

Meta: For context, why are filenames *not* required for `text.Failure`s? Shouldn't every lint failure should have a corresponding file?

---

protoc parsing can be an issue, and sometimes you don’t get a filename, see the regex parsing in the protoc package. I’m going to think more about this tomorrow and see if I can come up with better guarantees.

---

One thing I might do is fail the prototool call with a system error if harbormaster output can not be created for this reason. So high level, if you specify the harbormaster flag, and a failure does not have a filename, you won’t get a text.Failure, you’ll get a system error at the top level. Thoughts?

---

sounds good to me!

---

Nice, sounds good 👍

---

Done